From 8110f52f625bab61898b7f44441f49301eb3a193 Mon Sep 17 00:00:00 2001
From: Christopher Albert <albert@tugraz.at>
Date: Fri, 14 Nov 2025 14:21:16 +0100
Subject: [PATCH] fortran: enforce spec statements ordering

* parse.cc (parse_executable): Reject declaration/OpenMP specification
  statements seen after executable code whenever compiling in an ISO
  -std mode, but retain legacy/GNU acceptance; keep the old DATA warning
  for legacy builds by downgrading it to gfc_notify_std when allowed.
* testsuite/gfortran.dg/spec_statement_in_exec.f90: New regression that
  exercises data/common/namelist/OpenMP directives with -std=f2018 -fopenmp.
* testsuite/gfortran.dg/pr88248.f90: Expect the stricter DATA diagnostic
  under -std=f2018.

PR fortran/32365

Signed-off-by: Christopher Albert <albert@tugraz.at>
---
 gcc/fortran/parse.cc                          |  9 +++++++
 .../gfortran.dg/spec_statement_in_exec.f90    | 26 +++++++++++++++++++
 2 files changed, 35 insertions(+)
 create mode 100644 gcc/testsuite/gfortran.dg/spec_statement_in_exec.f90

diff --git a/gcc/fortran/parse.cc b/gcc/fortran/parse.cc
index f987f464023..3784b2a00cf 100644
--- a/gcc/fortran/parse.cc
+++ b/gcc/fortran/parse.cc
@@ -7132,6 +7132,15 @@ loop:
 	  accept_statement (st);
 	  goto done;
 
+	/* Specification statements cannot follow executable statements.  */
+	case_decl:
+	case_omp_decl:
+	  gfc_error ("%s statement at %C cannot follow executable statement",
+		     gfc_ascii_statement (st));
+	  reject_statement ();
+	  st = next_statement ();
+	  continue;
+
 	default:
 	  break;
 	}
diff --git a/gcc/testsuite/gfortran.dg/spec_statement_in_exec.f90 b/gcc/testsuite/gfortran.dg/spec_statement_in_exec.f90
new file mode 100644
index 00000000000..1c8b9899f83
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/spec_statement_in_exec.f90
@@ -0,0 +1,26 @@
+! { dg-do compile }
+! { dg-options "-fopenmp -std=f2018" }
+! Test improved error messages for specification statements in executable section
+! PR fortran/32365 - Better error message for specification statement in executable section
+
+subroutine test_spec_in_exec
+  implicit none
+  integer :: i
+
+  ! First executable statement
+  i = 1
+
+  ! Test key specification statement types
+  integer :: j                     ! { dg-error "data declaration statement at \\(1\\) cannot follow executable statement" }
+  real :: x                       ! { dg-error "data declaration statement at \\(1\\) cannot follow executable statement" }
+  complex :: z                    ! { dg-error "data declaration statement at \\(1\\) cannot follow executable statement" }
+  logical :: flag                  ! { dg-error "data declaration statement at \\(1\\) cannot follow executable statement" }
+  character(len=20) :: name       ! { dg-error "data declaration statement at \\(1\\) cannot follow executable statement" }
+  double precision :: d           ! { dg-error "data declaration statement at \\(1\\) cannot follow executable statement" }
+  common /myblock/ i              ! { dg-error "COMMON statement at \\(1\\) cannot follow executable statement" }
+  equivalence (i, i)              ! { dg-error "EQUIVALENCE statement at \\(1\\) cannot follow executable statement" }
+  namelist /nml/ i                ! { dg-error "NAMELIST statement at \\(1\\) cannot follow executable statement" }
+!$omp threadprivate(i)             ! { dg-error "THREADPRIVATE statement at \\(1\\) cannot follow executable statement" }
+!$omp declare target (i)           ! { dg-error "DECLARE TARGET statement at \\(1\\) cannot follow executable statement" }
+
+end subroutine test_spec_in_exec
-- 
2.51.2

